### PR TITLE
DDF 2874 - Add better support for geometry attributes, refactor property based views, fix issue with enumeration validator

### PR DESCRIPTION
**Pull request by andrewkfiedler**
#### What does this PR do?
 - Adds better support for geometry attributes in the UI.  They now appear in the table and in the details pane as their wkt form.  Each geometry on the metacard is now displayed on the map as well.
 - Adds support for editing geometry attributes in the UI.  Validation is run on the user values to help point out errors before the user clicks save.  One checks if the wkt is malformed, the other checks whether or not the coordinate values make sense in lon / lat ordering.
 - Refactors property based views to move towards updating the property value directly rather than waiting for the save event to collect values.  See listenForChange on the various property views.  Updated areas where people were listening to the views rather than the model to now listen to the model.
 - Adds better support for validation to the input views.  In the future we should try to add better support for the various validation rules that admins can write, so users can immediately test whether or not their changes will go through.  The enum input has already been updated to take advantage of this.  Basically, input views can have validation of their own, and administrators can add on validation as well.
 - Fixes enumeration validator to not throw a class cast exception on number based enums.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@andrewkfiedler
@jlcsmith 
@rzwiefel 
@bdeining 

#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-2874

### CODE REVIEW

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                    properties.put(attribute.getName(), value);
+                }
+            }
+        }
+
+        properties.put(METACARD_TYPE_PROPERTY_KEY,
+                metacard.getMetacardType()
+                        .getName());
+
+        if (metacard.getSourceId() != null && !"".equals(metacard.getSourceId())) {
```
Comment (bdeining) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java]: You can use `StringUtils.isNotBlank(metacard.getSourceId())` here to simplify the logic

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/
+
+ -->
+<blueprint xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0"
+		>
```
Comment (bdeining) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: odd formatting

File: catalog/ui/catalog-ui-search/src/main/webapp/component/input/bulk/input-bulk.view.js [hunk 3]
```diff
@@ -35,11 +35,32 @@ define([
         },
         events: {
         },
+        listenForChange: function(){
+           // if (!this.model.isHomogeneous()){
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/input/bulk/input-bulk.view.js]: remove?

File: catalog/ui/catalog-ui-search/src/main/webapp/component/input/bulk/input-bulk.view.js [hunk 4]
```diff
@@ -35,11 +35,32 @@ define([
...
+                            this.model.setValue(value);
+                        break;
+                    }
+                    this.handleChange();
+                });
+                this.listenTo(this.otherInput.currentView.model, 'change:value', function(){
+                    this.model.setValue(this.otherInput.currentView.model.getValue());
+                    this.handleChange();
+                });
+         //   }
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/input/bulk/input-bulk.view.js]: remove?

File: catalog/ui/catalog-ui-search/src/main/webapp/component/input/bulk/input-bulk.view.js [hunk 5]
```diff
@@ -35,11 +35,32 @@ define([
...
+           // if (!this.model.isHomogeneous()){
+                this.listenTo(this.enumRegion.currentView.model, 'change:value', function(){
+                    var value = this.enumRegion.currentView.model.get('value')[0];
+                    switch(value){
+                        case 'bulkDefault':
+                            this.model.revert();
+                        break;
+                        case 'bulkCustom':
+                            this.model.setValue(this.otherInput.currentView.model.getValue());
+                            break;
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/input/bulk/input-bulk.view.js]: formatting is odd here

File: catalog/ui/catalog-ui-search/src/main/webapp/component/input/input.view.js [hunk 6]
```diff
@@ -52,6 +52,22 @@ define([
...
+            this.$el.on('change keyup input', function(){
+                this.model.set('value', this.getCurrentValue());
+                this.handleValidation();
+            }.bind(this));
+        },
+        handleValidation: function(){
+            this.$el.toggleClass('has-validation-issues', !this.isValid());
+        },
+        isValid: function(){
+            return true; //overwrite on a per input basis
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/input/input.view.js]: Should this be always true?  Curious what is overwritten

File: catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.js [hunk 7]
```diff
@@ -24,24 +26,100 @@ define([
...
             this.setDefaultLabel();
+            this.listenTo(this, 'change:value', this.updateHasChanged);
         },
         setDefaultLabel: function(){
             if (!this.get('label')){
                 this.set('label', this.get('id'));
             }
         },
+        hasChanged: function(){
+            if (!this.isHomogeneous()){
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.js]: remove?

File: catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.js [hunk 8]
```diff
@@ -69,13 +147,15 @@ define([
...
         },
-        save: function(value){
-            this.set('value', value);
-            this.trigger('change:value');
+        revert: function(){
+            this.set({
+                hasChanged: false,
+                value: this.getInitialValue()
+            });
+          //  this.trigger('change:value');
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.js]: remove?

File: catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.view.js [hunk 9]
```diff
@@ -82,9 +83,10 @@ define([
             this.$el.addClass('has-limited-width');
         },
         revert: function(){
+            this.model.revert();
             this.onBeforeShow();
-            this.$el.trigger('change');
-            this.$el.trigger('revert');
+        //    this.$el.trigger('change');
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.view.js]: remove?

File: catalog/ui/catalog-ui-search/src/main/webapp/component/value/value.collection.view.js [hunk 10]
```diff
@@ -26,44 +26,34 @@ define([
     return Marionette.CollectionView.extend({
         childView: ValueView,
         tagName: CustomElements.register('value-collection'),
-        onRender: function () {
+        collectionEvents: {
+            'change:value': 'updateProperty',
+          //  'change:hasChanged': 'updatePropertyHasChanged',
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/value/value.collection.view.js]: remove?

File: catalog/ui/catalog-ui-search/src/main/webapp/component/visualization/maps/cluster.collection.view.js [hunk 11]
```diff
@@ -34,7 +35,7 @@ var ClusterCollectionView = Marionette.CollectionView.extend({
         this.selectionInterface = options.selectionInterface || this.selectionInterface;
         this.options.map.onLeftClick(this.onMapLeftClick.bind(this));
         this.options.map.onMouseMove(this.handleMapHover.bind(this));
-        this.listenTo(this.selectionInterface.getActiveSearchResults(), 'reset', this.handleResultsChange);
+        //this.listenTo(this.selectionInterface.getActiveSearchResults(), 'reset', this.handleResultsChange);
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/visualization/maps/cluster.collection.view.js]: remove?

File: catalog/ui/catalog-ui-search/src/main/webapp/component/input/input.view.js (hunk 6 above)
Comment (andrewkfiedler) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/input/input.view.js]: The other inputs that people make overwrite this function.  By default, it returns true, but a special input (such as the input-enum) can do it's own validation and return true or false.

File: catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.js (hunk 7 above)
Comment (andrewkfiedler) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.js]: This one should stay.

File: catalog/transformer/catalog-transformer-propertyjson-metacard/pom.xml [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+                        <configuration>
+                            <haltOnFailure>true</haltOnFailure>
+                            <rules>
+                                <rule>
+                                    <element>BUNDLE</element>
+                                    <limits>
+                                        <limit>
+                                            <counter>INSTRUCTION</counter>
+                                            <value>COVEREDRATIO</value>
+                                            <minimum>0.0</minimum>
```
Comment (andrewkfiedler) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/pom.xml]: These should be set to the actual values

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+    protected static final MimeType DEFAULT_MIME_TYPE = new MimeType();
+
+    public static final String ID = "propertyjson";
+
+    static {
+        try {
+            DEFAULT_MIME_TYPE.setPrimaryType("application");
+            DEFAULT_MIME_TYPE.setSubType("json");
+        } catch (MimeTypeParseException e) {
+            LOGGER.info("Failure creating MIME type", e);
```
Comment (andrewkfiedler) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java]: make sure info is okay

Comment (andrewkfiedler) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java]: TODO(Ryan) - Maybe replace with constants so we don't have to deal with exception at all

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+            throw new ExceptionInInitializerError(e);
+        }
+        DATE_FORMAT.setTimeZone(TimeZone.getTimeZone("GMT"));
+    }
+
+    public static Map<String, Object> convertToJSON(Metacard metacard) throws CatalogTransformerException{
```
Comment (andrewkfiedler) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java]: need format

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+        case INTEGER:
+        case SHORT:
+            return value;
+        case STRING:
+        case XML:
+        case FLOAT:
+        case GEOMETRY:
+            return value.toString();
+        case OBJECT:
+        default:
```
Comment (andrewkfiedler) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java]: explicitly comment fallthrough

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/test/groovy/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformerSpec.groovy [hunk 16]
```diff
@@ -0,0 +1,81 @@
+package ddf.catalog.transformer.metacard.propertyjson
```
Comment (andrewkfiedler) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/test/groovy/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformerSpec.groovy]: Add header here

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlResult.java [hunk 17]
```diff
@@ -229,7 +229,7 @@ private Double normalizeDistance(Result result, Query query, FilterAdapter filte
                 }
             }
 
-            geoJson = GeoJsonMetacardTransformer.convertToJSON(resultMetacard);
+            geoJson = PropertyJsonMetacardTransformer.convertToJSON(resultMetacard);
```
Comment (andrewkfiedler) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlResult.java]: Need to add the dontInclude attributes

File: catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.collection.view.js [hunk 18]
```diff
@@ -219,7 +222,17 @@ define([
                     });
                     return false;
                 }
-            }).sort();
+            }).sort(function(a, b){
+                a = metacardDefinitions.getLabel(a);
```
Comment (andrewkfiedler) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/property/property.collection.view.js]: `toLowerCase()`

File: catalog/ui/catalog-ui-search/src/main/webapp/component/result-filter/result-filter.less [hunk 19]
```diff
@@ -2,6 +2,10 @@
   display: block;
   padding: 10px;
 
+  @{customElementNamespace}dropdown.is-simpleDropdown .dropdown-text {
+    text-overflow: clip;
```
Comment (andrewkfiedler) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/result-filter/result-filter.less]: review

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java (hunk 15 above)
Comment (rzwiefel) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java]: nvm

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 2 above)
Comment (rzwiefel) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: fixed

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/test/groovy/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformerSpec.groovy (hunk 16 above)
Comment (rzwiefel) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/test/groovy/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformerSpec.groovy]: done

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlResult.java (hunk 17 above)
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlResult.java]: done

**Comment by jlcsmith** [comment]
If any of the geometry attributes are hidden in the UI config, do we want them to not be displayed on the map in addition to the text views?

**Comment by andrewkfiedler** [comment]
build now

**Comment by jlcsmith** [comment]
build now

**Comment by jlcsmith** [comment]
Verified
- OSX / Windows
- Chrome/FF
- 2D/3D map
- Clustering
- Editing location including malformed WKT
- Ability to copy location

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java (hunk 13 above)
Comment (bdeining) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java]: What about logging at debug?  I could see a situation where this is too verbose on multiple failures

File: catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+    @Nullable
+    private static Object convertValue(Serializable value, AttributeType.AttributeFormat format)
+            throws CatalogTransformerException {
+        if (value == null) {
+            return null;
+        }
+
+        switch (format) {
+        case DATE:
+            // Creating date format instance each time is inefficient, however
```
Comment (bdeining) [inline catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java]: Have you considered https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/time/FastDateFormat.html?

**Pull request merged**